### PR TITLE
Catchup at

### PR DESCRIPTION
**Pull request by vogel**
Changes way that catchup is done (from recent/minimal/complete to catchup-count-ledgers-to-ledger, with auto detection if applying buckets or transaction is needed), then adds few simple functions to main.cpp.

### CODE REVIEW

File: src/catchup/CatchupConfiguration.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// checkpoint.
+class CatchupConfiguration
+{
+  public:
+    static const uint32_t CURRENT = 0;
+
+    /**
+     * Preconditions:
+     * * lastClosedLedger > 0
+     * * toLedger == 0 || toLedger >= lastClosedLedger
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupConfiguration.h]: I think you want `toLedger == CatchupConfiguration::CURRENT` not `0`

Comment (MonsieurNicolas) [reply src/catchup/CatchupConfiguration.h]: now that I think of it, you could also have avoided the use of magic values it would have been better, maybe by having two constructors, and passing CatchupConfiguration early on (right now it's constructed in startCatchup)

File: src/catchup/CatchupManager.h [hunk 2]
```diff
@@ -37,49 +36,15 @@ enum class VerifyLedgerMode
...
-    // to the next scheduled checkpoint boundary, instead catching up to a
-    // checkpoint presumed to have been made at `initLedger` (i.e. with
-    // checkpoint ledger number equal to initLedger-1). This 'manual' catchup
-    // mode exists to support catching-up to manually created checkpoints.
-    virtual void catchupHistory(uint32_t initLedger, CatchupMode mode,
-                                CatchupWork::ProgressHandler handler,
-                                bool manualCatchup = false) = 0;
+    // Run catchup with given configuration and verify mode.
+    virtual void catchupHistory(CatchupConfiguration catchupConfiguration,
+                                VerifyLedgerMode verifyMode,
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupManager.h]: any reason to not incorporate verifyMode with CatchupConfiguration, I don't see why we keep it as a separate field

File: src/catchup/CatchupRange.cpp [hunk 3]
```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
...
 }
+
+CatchupRange
+makeCatchupRange(ResolvedCatchupConfiguration const& configuration,
+                 HistoryManager const& historyManager)
+{
+    auto toCheckpoint =
+        historyManager.nextCheckpointLedger(configuration.toLedger() + 1) - 1;
+
+    auto firstNotKnownCheckpoint = [&]() {
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: why a lambda here?

File: src/catchup/CatchupRange.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#include "catchup/CatchupRange.h"
+#include <utility>
+
+namespace stellar
+{
+
+CatchupRange::CatchupRange(CheckpointRange range, uint32_t stopAt,
+                           bool applyBuckets)
+    : mRange{std::move(range)}, mStopAt{stopAt}, mApplyBuckets{applyBuckets}
+{
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: no validation for mStopAt?

File: src/catchup/CatchupRange.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+// equivalent to old CATCHUP_MIMIMAL.
+//
+// For old CATCHUP_COMPLETE applyBuckets() is always false, as CATCHUP_COMPLETE
+// only applies transactions.
+//
+// For old CATCHUP_RECENT value of applyBuckets() depends on value of last
+// closed ledger - if CATCHUP_RECENT would go before it, it behaves exactly as
+// CATCHUP_COMPLETE. If not, applyBuckets() is true and miminal catchup will
+// be done first.
+class CatchupRange
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.h]: This class doesn't seem to make sense to me. Especially as you are introducing a CatchupConfiguration in the next commit.

File: src/catchup/CatchupRange.cpp [hunk 6]
```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
...
+        {
+            return nextCheckpoint + historyManager.getCheckpointFrequency();
+        }
+        return nextCheckpoint;
+    }();
+
+    auto catchupFromLedger =
+        configuration.toLedger() > configuration.count()
+            ? configuration.toLedger() - configuration.count()
+            : 0;
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: what does 0 mean here?

File: src/catchup/CatchupRange.cpp [hunk 7]
```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
...
+    auto catchupFromCheckpoint =
+        historyManager.nextCheckpointLedger(catchupFromLedger) - 1;
+
+    // ensure that we have at least configuration.count() ledgers if
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: this needs better comments/variable names, what is really happening here is that we're attempting to go to the previous checkpoint if the one currently computed brings us before catchupFromLedger (the highest ledger we should be starting from)

File: src/catchup/CatchupRange.cpp [hunk 8]
```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
...
+    // possible
+    if (catchupFromLedger < catchupFromCheckpoint &&
+        catchupFromCheckpoint > historyManager.getCheckpointFrequency())
+    {
+        catchupFromCheckpoint -= historyManager.getCheckpointFrequency();
+    }
+
+    auto fromCheckpoint =
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: is this correct (computation of fromCheckpoint)?
It seems that it should be catchupFromCheckpoint if we have computed it successfully otherwise default to firstNotKnownCheckpoint : I think there is an edge case when firstNotKnownCheckpoint  is in the future (which will cause the local node to wait for the next checkpoint instead of downloading data from the previous checkpoint).

File: src/catchup/CatchupRange.cpp [hunk 9]
```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
...
+        std::max(catchupFromCheckpoint, firstNotKnownCheckpoint);
+
+    auto checkpointRange = CheckpointRange{fromCheckpoint, toCheckpoint};
+    auto applyBuckets = fromCheckpoint > firstNotKnownCheckpoint;
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: I am not sure I follow how `applyBuckets` is computed, shouldn't it be related not to the firstNotKnownCheckpoint but instead just the LCL, ie if we can't replay tx directly, we have to jump to a place in history such that we have at least `configuration.count()` ledgers worth of transactions being replayed?

File: src/catchup/CatchupWork.cpp [hunk 10]
```diff
@@ -3,55 +3,230 @@
...
+    auto toLedger = mCatchupConfiguration.toLedger() == 0
+                        ? "CURRENT"
+                        : std::to_string(mCatchupConfiguration.toLedger());
+    CLOG(INFO, "History") << "Starting catchup with configuration:\n"
+                          << "  lastClosedLedger: "
+                          << mCatchupConfiguration.lastClosedLedger() << "\n"
+                          << "  toLedger: " << toLedger << "\n"
+                          << "  count: " << mCatchupConfiguration.count();
+
+    auto toCheckpoint =
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupWork.cpp]: this seems wrong: the (complex) logic to compute what to download etc is in `makeCatchupRange` but that code is only triggered after kicking off `GetHistoryArchiveStateWork`

Comment (MonsieurNicolas) [reply src/catchup/CatchupWork.cpp]: oh nm, this is to fetch the remote state from archive

File: src/catchup/ResolvedCatchupConfiguration.h [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+#include <cstdint>
+
+namespace stellar
+{
+
+class CatchupConfiguration;
+
+// Version of CatchupConfiguration class where value of toLedger is never
+// CatchupConfiguration::CURRENT - it is always real ledger numer.
+class ResolvedCatchupConfiguration
```
Comment (MonsieurNicolas) [inline src/catchup/ResolvedCatchupConfiguration.h]: This seems overkill, why not pass around a `CatchupConfiguration` and throw if it has a value that we don't expect. More code is not better.

File: src/main/CommandHandler.cpp [hunk 12]
```diff
@@ -534,12 +535,13 @@ CommandHandler::catchup(std::string const& params, std::string& retStr)
...
-    mApp.getLedgerManager().startCatchUp(ledger, mode, true);
+    mApp.getLedgerManager().startCatchUp(ledger, count,
+        VerifyLedgerMode::DO_NOT_VERIFY_BUFFERED_LEDGERS);
     retStr = (std::string("Started catchup to ledger ") +
               std::to_string(ledger) + std::string(" in mode ") +
-              std::string(mode == CatchupManager::CATCHUP_COMPLETE
+              std::string(count == std::numeric_limits<uint32_t>::max()
                               ? "CATCHUP_COMPLETE"
-                              : (mode == CatchupManager::CATCHUP_RECENT
+                              : (count > 0
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: `count != 0` would be more accurate as it's an `uint32_t`

File: src/main/main.cpp [hunk 13]
```diff
@@ -502,8 +557,20 @@ main(int argc, char* const* argv)
...
-            catchupComplete = true;
+            doCatchupComplete = true;
+            break;
+        case OPT_CATCHUP_RECENT:
+            doCatchupRecent = true;
+            catchupRecentCount = atol(optarg);
+            break;
+        case OPT_CATCHUP_TO:
+            doCatchupTo = true;
+            catchupToTarget = atol(optarg);
```
Comment (MonsieurNicolas) [inline src/main/main.cpp]: given that this has the potential to blow up people's databases, we should not use `atol` to parse parameters. Also `0` should not be allowed as a magic number and people should use `current` or `latest` instead.

File: src/catchup/CatchupConfiguration.h (hunk 1 above)
Comment (vogel) [reply src/catchup/CatchupConfiguration.h]: I create CatchupConfiguration in startCatchup because of the lastClosedLedger that is available inside LedgerManagerImpl. Previous version passed CatchupConfiguration to startCatchup, but there was a lot of repetition with setting the lastClosedLedger in it - it always used the same value from LedgerManagerImpl.

So I moved that and made setting lastClosedLedger responsibility of LedgerManagerImpl, so there is no possibility of passing there an invalid value.

File: src/catchup/CatchupRange.cpp (hunk 3 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: no good reason, can be replaced by simple ?: statement in last line

File: src/catchup/CatchupRange.cpp (hunk 4 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: Not sure how to do that, It should be in range of [range.last() - checkpoint_size + 1...range.last()] but checkpoint_size is user-configurable. Also passing HistoryManager to that constructor does not seem to be a good idea.

Do you have any suggestions?

File: src/catchup/CatchupRange.h (hunk 5 above)
Comment (vogel) [reply src/catchup/CatchupRange.h]: CatchupRange life time is as follows:
* CatchupConfiguration is created with, LCL, catchup-to and catchup-count values
* ResolvedCatchupConfiguration is created that replaces catchup-to to a value that is not CatchupConfiguration::CURRENT (in case of command line catchups)
* CatchupRange is computed from ResolvedCatchupConfiguration and local state of ledger

So CatchupRange is internal class to CatchupWork as its creation requires many information that is just not available at the moment of CatchupConfiguration creation (mainly what is the last ledger on network).

Also CatchupConfiguration shows what we want to do (get last X ledgers to ledger Y or to last ledger) and CatchupRange shows how to do that (apply buckets at checkpoint F and then apply transactions to checkpoint L).

File: src/catchup/CatchupRange.cpp (hunk 6 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: It means ledger number zero. Probably could use 1 or 2 here as well.

File: src/catchup/CatchupRange.cpp (hunk 7 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: I would phrase it as:

`we're attempting to go to the previous checkpoint if the one currently computed would make us apply less ledgers than configuration.count() `

File: src/catchup/ResolvedCatchupConfiguration.h (hunk 11 above)
Comment (vogel) [reply src/catchup/ResolvedCatchupConfiguration.h]: Possibly. I think that this code makes it compile-time error to pass wrong value (CatchupConfiguration::CURRENT) to a code that does not expects it. I can remove that if you wish and just add asserts to all code that currently receives ResolvedCatchupConfiguration.

File: src/main/main.cpp (hunk 13 above)
Comment (vogel) [reply src/main/main.cpp]: what should we use instead? `atoi` or do you mean something else?

File: src/main/CommandHandler.cpp (hunk 12 above)
Comment (vogel) [reply src/main/CommandHandler.cpp]: ok

File: src/catchup/CatchupRange.cpp (hunk 9 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: Well, that is complicated.

firstNotKnownCheckpoint is related to LCL.

If we have for example LCL = 1022, then firstNotKnownCheckpoint is 1023 (as this is the first checkpoint that is not complety known).

Then fromCheckpoint value will be at least 1023 (thanks to the std::max) function. If it is equal to 1023 then we don't need to apply buckets - we will get transactions for checkpoints from 1023 to whatever toCheckpoint is. Transactions for checkpoint 1023 are: all from 958 to 1023, so we will just apply all of them (and skip all but last as implemented in ApplyLedgerChainWork).

If however fromCheckpoint is bigger, like 1087, then we need to apply buckets at 1087, as replaying ledgers for that checkpoint (1024 to 1087) would not work, with 1023 missing in our database.

File: src/catchup/CatchupRange.cpp (hunk 4 above)
Comment (MonsieurNicolas) [reply src/catchup/CatchupRange.cpp]: it sounds like having checkpoint member of `CheckpointRange` would solve this. In the previous PR I thought we would only use it for validation of `CheckpointRange`, but it makes sense to remember it in that structure

File: src/catchup/CatchupRange.h (hunk 5 above)
Comment (MonsieurNicolas) [reply src/catchup/CatchupRange.h]: What I was saying is that CatchupWork passed down to the various *Work doesn't make much sense and creates weird couplings: the old interfaces were only taking what the *Work needed, now they take this object type as if they cared about it, but they don't. For example `DownloadAndApplyTransactionsWork` doesn't care about `applyBuckets` or why the range is what it is, it just needs a start and an end.

File: src/catchup/ResolvedCatchupConfiguration.h (hunk 11 above)
Comment (MonsieurNicolas) [reply src/catchup/ResolvedCatchupConfiguration.h]: yeah it's common practice to have argument validation and throw a `bad_argument` exception

File: src/main/main.cpp (hunk 13 above)
Comment (MonsieurNicolas) [reply src/main/main.cpp]: you can use `std::stoi` (or other flavor like `std::stoul`), and make sure to use the optional `pos` parameter to ensure that the entire string is consumed.

**Comment by MonsieurNicolas** [comment]
I think that given the significant changes this PR should also contain some of the changes we discussed before (which may uncover other problems with catchup that we didn't look into yet):
* addition of metrics for the various `*Work` classes (in the context of catchup)
* we can then update tests such that we check those metrics. In particular we can verify that the various catchup related tests are performing the work we expect them to do (and not more).
   * number of buckets downloaded and/or applied
   * number of tx downloaded and/or applied
   * number of ledger headers downloaded
   * number of times the various `verify` functions were called

File: src/catchup/ApplyBucketsWork.cpp [hunk 14]
```diff
@@ -117,19 +126,28 @@ ApplyBucketsWork::onStart()
...
+        else if (mCurrApplicator && *mCurrApplicator)
+        {
+            mCurrApplicator->advance();
+        }
     }
-    else if (mCurrApplicator && *mCurrApplicator)
+    catch (...)
     {
-        mCurrApplicator->advance();
+        mBucketApplyFailure.Mark();
```
Comment (MonsieurNicolas) [inline src/catchup/ApplyBucketsWork.cpp]: this probably belongs in the `OnFailure` family of handlers: from what I understand `Work` methods are not supposed to throw

File: src/catchup/ApplyBucketsWork.cpp [hunk 15]
```diff
@@ -31,12 +31,12 @@ ApplyBucketsWork::ApplyBucketsWork(
     , mFirstVerified(firstVerified)
     , mApplying(false)
     , mLevel(BucketList::kNumLevels - 1)
-    , mBucketApplyStart(app.getMetrics().NewMeter(
```
Comment (MonsieurNicolas) [inline src/catchup/ApplyBucketsWork.cpp]: from what I see you replaced meters by counters, which seems to be more of a downgrade as a meter tracks both count and rate, where as a counter only tracks count.
If you were looking at getting access to the counter stored inside a meter, see [`medida::Meter::count`](http://dln.github.io/medida/api/classmedida_1_1Meter.html#afb53f91f1ef6a53e33c82d501353443b)

File: src/catchup/ApplyBucketsWork.cpp [hunk 16]
```diff
@@ -126,28 +126,20 @@ ApplyBucketsWork::onStart()
...
         mApplying = true;
-        mBucketApplyStart.Mark();
+        mBucketApplyStart.inc();
     }
 }
 
 void
 ApplyBucketsWork::onRun()
 {
-    try
```
Comment (MonsieurNicolas) [inline src/catchup/ApplyBucketsWork.cpp]: This doesn't belong to this commit

File: src/history/HistoryTests.cpp [hunk 17]
```diff
@@ -601,19 +709,157 @@ HistoryTests::catchupApplication(uint32_t initLedger, uint32_t count,
...
+    auto& verifyLedgerChainFailure = app->getMetrics().NewCounter(
+        {"history", "verify-ledger-chain", "failure"});
+    auto& verifyLedgerChainFailureEnd = app->getMetrics().NewCounter(
+        {"history", "verify-ledger-chain", "failure-end"});
+
+    auto ledgersVerified = verifyLedgerSuccess.count();
+    REQUIRE(verifyLedgerSuccessOld.count() == 0);
+    REQUIRE(verifyLedgerFailureOvershot.count() == 0);
+    REQUIRE(verifyLedgerFailureLink.count() == 0);
+    //     REQUIRE(verifyLedgerChainFailure.count() == 0);
```
Comment (MonsieurNicolas) [inline src/history/HistoryTests.cpp]: commented out?

File: src/history/HistoryTests.cpp [hunk 18]
```diff
@@ -662,8 +908,10 @@ TEST_CASE_METHOD(HistoryTests, "Full history catchup",
 
     std::vector<Application::pointer> apps;
 
-    std::vector<uint32_t> counts = {0, std::numeric_limits<uint32_t>::max(),
-                                    60};
+    std::vector<std::tuple<uint32_t, CatchupMetrics>> cfgs = {
+        {0, {1, 1, 64, 4, 4, 1, 0}},
```
Comment (MonsieurNicolas) [inline src/history/HistoryTests.cpp]: I think we should not hardcode numbers like this as it's impossible to know if they are correct or how to change them in the future.
Instead we should have them computed using formulas using expected behavior. For example: a bucketlist with 3 non-empty buckets should trigger 3 downloads

File: src/history/HistoryTests.cpp [hunk 19]
```diff
@@ -96,6 +96,105 @@ class TmpDirConfigurator : public Configurator
     }
 };
 
+struct CatchupMetrics
```
Comment (MonsieurNicolas) [inline src/history/HistoryTests.cpp]: This is the type of stuff I was talking about. Nice.

At the same time, maybe we want to break this up so that it maps to Work classes? ie, if I have a CatchupWork that spawns 2 types of Work, I don't need to flatten out the metrics in one structure but instead have metrics for each Work implementation.

File: src/history/HistoryTests.cpp (hunk 18 above)
Comment (MonsieurNicolas) [reply src/history/HistoryTests.cpp]: Now that I think of it, if you were breaking down the "expected metrics" classes such that they map 1:1 to each *Work, then you should be able to construct each of those with high level parameters and the numbers broken up by Work item should be self-explanatory.

File: src/catchup/ApplyBucketsWork.cpp (hunk 15 above)
Comment (vogel) [reply src/catchup/ApplyBucketsWork.cpp]: I've changed that so I can reset (counters can be reset, I don't see that option for meters). I need resets for tests.

File: src/catchup/ApplyBucketsWork.cpp (hunk 16 above)
Comment (vogel) [reply src/catchup/ApplyBucketsWork.cpp]: Right. I'll update that when I'll be doing rebases when it is accepted.

File: src/history/HistoryTests.cpp (hunk 18 above)
Comment (vogel) [reply src/history/HistoryTests.cpp]: I can create them form CatchupRange class - is that what you mean?

File: src/catchup/ApplyBucketsWork.cpp (hunk 15 above)
Comment (MonsieurNicolas) [reply src/catchup/ApplyBucketsWork.cpp]: Oh I see - it seems better to just add a `clear` method to `Meter`. It seems like a legitimate change to the library that can also be submitted upstream if you feel like it (but for now you can just make the change in our fork as that project seems to be abandoned at the moment).

File: src/catchup/CatchupRange.cpp (hunk 6 above)
Comment (MonsieurNicolas) [reply src/catchup/CatchupRange.cpp]: yeah I'd rather not use 0 as 0 is not a valid ledger number

File: src/catchup/CatchupRangeTests.cpp [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+#include <lib/catch.hpp>
+#include <lib/util/format.h>
+
+using namespace stellar;
+
+auto max = std::numeric_limits<uint32_t>::max();
+namespace stellar
+{
+std::vector<std::pair<CatchupConfiguration, CatchupRange>> gCatchupRangeCases{
+    {{1, 2, 0}, {{{63, 63, 64}, 2}, false}},
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRangeTests.cpp]: Add a comment on how you came up with this test data.

File: src/catchup/CatchupWork.cpp [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+CatchupWork::verifyLedgers(CheckpointRange const& range)
+{
+    if (mVerifyLedgersWork)
+    {
+        assert(mVerifyLedgersWork->getState() == WORK_SUCCESS);
+        return false;
+    }
+
+    CLOG(INFO, "History")
+        << "Catchup veryfing ledger chain for checkpointRange ["
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupWork.cpp]: typo `veryfing`

File: src/history/HistoryTests.cpp (hunk 18 above)
Comment (MonsieurNicolas) [reply src/history/HistoryTests.cpp]: I see that you did something closer to what I meant in `split CatchupMetrics into smaller structs` later, but it doesn't look like the new proposed split is enough to make those numbers less cryptic.

I think that the inputs specified at the test level should not be those numbers (in most cases), but instead be describing the high level steps that describes the scenario being tested.

Maybe a little more detail on what I am thinking can help.

Looking at `CatchupWork`, it spawns the following Work:
 * mGetHistoryArchiveStateWork;
 * mDownloadLedgersWork;
 * mVerifyLedgersWork;
 * mGetBucketsHistoryArchiveStateWork;
 * mDownloadBucketsWork;
 * mApplyBucketsWork;
 * mDownloadTransactionsWork;
 * mApplyTransactionsWork;

So what you need is a way to translate a given scenario into what it means for those spun up work items which then can translated into the metrics you have.

Example:
Scenario being tested - catchup complete to ledger #64 (a checkpoint), from a blank node - no failures.
we expect in this scenario to have the something like the following sequence:
 * mGetHistoryArchiveStateWork: archive queried once
 * mDownloadLedgersWork: should download the files to fetch ledger headers 3..64
 * mVerifyLedgersWork: should run once
 * mGetBucketsHistoryArchiveStateWork: should do nothing
 * mDownloadBucketsWork: should do nothing
 * mApplyBucketsWork: should do nothing
 * mDownloadTransactionsWork: should download files to cover ledger transitions 2->3, 3->4, …, 63->64
 * mApplyTransactionsWork: should close 62 ledgers
	
And from this, I think you should be able to compute the expected numbers.

File: src/catchup/CatchupRange.cpp [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+CatchupRange
+makeCatchupRange(CatchupConfiguration const& configuration,
+                 HistoryManager const& historyManager)
+{
+    assert(configuration.toLedger() != CatchupConfiguration::CURRENT);
+
+    auto checkpointContainingToLedger =
+        historyManager.nextCheckpointLedger(configuration.toLedger() + 1) - 1;
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupRange.cpp]: I was expecting `checkpointContainingToLedger` to just be `historyManager.nextCheckpointLedger(configuration.toLedger())` per the formula & comment on what `nextCheckpointLedger` does.

File: src/catchup/ApplyLedgerChainRange.h [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+#include "ledger/CheckpointRange.h"
+
+namespace stellar
+{
+
+// Range of transaction checkpoints to apply.
+//
+// Transactions for checkpoins in checkpointRange() should be applied,
+// stoppint at ledger stopAt(). Ledger stopAt() must be in last checkpoint.
+class ApplyLedgerChainRange
```
Comment (MonsieurNicolas) [inline src/catchup/ApplyLedgerChainRange.h]: This code (in general, this class is part of the problem) is still quite hard to read as it still deals with too many types.

From what I see, we still have many range related classes:
 * CheckpointRange - (first, last, frequency)
 * ApplyLedgerChainRange - (mCheckpointRange, stopAt)
 * CatchupConfiguration - (lcl, to, count)
 * CatchupRange - (mApplyLedgerChain, applybuckets bool)

I don't think having all those objects makes sense as they are only a reflection of the state of how catchup is working, especially now that you have a centralized catchup implementation.

If you look at how catchup is implemented now.

CatchupWork, derives from BucketDownloadWork and spins up:
 * GetHistoryArchiveStateWork
 * BatchDownloadWork (checkpointrange )
 * VerifyLedgerChainWork (checkpointrange)
 * GetHistoryArchiveStateWork
 * DownloadBucketsWork
 * ApplyBucketsWork
 * BatchDownloadWork (checkpointrange)
 * ApplyLedgerChainWork(applyledgerchainrange)

From what I see, the only real benefit is to use CheckpointRange for work items that would otherwise fail. The only one that I can see that falls into this category is `CatchupWork` because of `BatchDownloadWork` 
 that has a strong dependency on checkpoint frequency.

The others (`VerifyLedgerChainWork` and `ApplyLedgerChainWork` ) should be able to work on an arbitrary range (the fact that it's lined up with checkpoints is only a concern for `CatchupWork`).

Now, as far as I can tell, the only reason you need all those types is that you want to ensure that the output of some calculation/work is properly used as the input to a different calculation/process at compile time but this is only a side effect of the way the code evolved: now that catchup is centralized you can have the code broken up in "steps" that can perform those computations on the fly using helper methods.

So here is what I am proposing:
* refactor the code to only use a minimum number of ledger number variables.
    * I think you only need to track the [`low`,`high`] being currently used by `CatchupWork` 
 (what is being downloaded/verified/applied etc) as well as the initial parameters (`toLedger` - that can be `CURRENT` and `count`)
    * a `CheckpointRange` should be computed in order to spin up `BatchDownloadWork` but is only used to update [low,high] at the `CatchupWork` level (it's basically a prep step before doing any work)
    * I do like that we have `CheckpointRange` as a parameter to BatchDownloadWork`
* all code should fall in place fairly easily as members of `CatchupWork`
* refactor the code such that `makeCatchupRange` becomes just an internal method to the `CatchupWork` class (or a few methods - some could be static so that you can continue to use the same tests that you wrote)

File: src/catchup/VerifyLedgerChainWork.cpp [hunk 24]
```diff
@@ -192,18 +215,18 @@ VerifyLedgerChainWork::onSuccess()
...
+                                  << mRange.last() << "] verified";
             return WORK_SUCCESS;
         }
 
         mCurrSeq += mApp.getHistoryManager().getCheckpointFrequency();
         return WORK_RUNNING;
     case HistoryManager::VERIFY_HASH_UNKNOWN_RECOVERABLE:
         CLOG(WARNING, "History")
-            << "Catchup material verification inconclusive, retrying";
+            << "Catchup material verification Markonclusive, retrying";
```
Comment (MonsieurNicolas) [inline src/catchup/VerifyLedgerChainWork.cpp]: keyboard spill `Markonclusive`

File: src/ledger/LedgerManagerImpl.cpp [hunk 25]
```diff
@@ -530,34 +538,34 @@ LedgerManagerImpl::historyCaughtup(asio::error_code const& ec,
...
         {
-            mLastClosedLedger = lastClosed;
-            CLOG(INFO, "Ledger") << "First phase of CATCHUP_RECENT done: "
-                                 << ledgerAbbrev(mLastClosedLedger);
-            mCurrentLedger = make_shared<LedgerHeaderFrame>(lastClosed);
-            return;
-        }
+            case CatchupWork::ProgressState::APPLIED_BUCKETS:
+            {
+                mCurrentLedger = make_shared<LedgerHeaderFrame>(lastClosed.header);
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerManagerImpl.cpp]: not necessarily for this PR (I see it was already like this), but it's a bit odd to have to update `mLastClosedLedger` here: I would have imagined that applying buckets would take care of it (just like replaying transactions indirectly updates it)

File: src/ledger/LedgerManagerImpl.cpp [hunk 26]
```diff
@@ -530,34 +538,34 @@ LedgerManagerImpl::historyCaughtup(asio::error_code const& ec,
...
+                storeCurrentLedger();
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerManagerImpl.cpp]: I didn't realize that we were not saving the state during catching up (internally to `CatchupWork`)... isn't that a potential source of problem if we only do partial catchup? Like if we apply buckets and crash/timeout when downloading transactions?

File: src/catchup/CatchupRange.cpp (hunk 22 above)
Comment (vogel) [reply src/catchup/CatchupRange.cpp]: It is not unfortunately.
if toLedger = 64 then nextCheckpointLedger = 64, but 64 is in checkpoint 127

I'll make new set of methods for such computation so it is more obvious what is happening.

Comment (MonsieurNicolas) [reply src/catchup/CatchupRange.cpp]: Oh, I see I looked at those helper methods in more detail it's because they operate on "current" instead of LCL, but what is published is LCL... so we end up with the weird +1/-1 here. Don't add new methods as they would only be used here from what I can tell.

**Comment by vogel** [comment]
Looks like I've broken a merge somehow. Let me try to fix that.

**Comment by vogel** [comment]
Looks fixed now.

File: src/catchup/CatchupWork.cpp [hunk 27]
```diff
@@ -0,0 +1,498 @@
...
+CatchupRange
+CatchupWork::makeCatchupRange(uint32_t lastClosedLedger,
+                              CatchupConfiguration const& configuration,
+                              HistoryManager const& historyManager)
+{
+    assert(lastClosedLedger > 0);
+    assert(configuration.toLedger() >= lastClosedLedger);
+    assert(configuration.toLedger() != CatchupConfiguration::CURRENT);
+
+    // smallest ledger that should be applied is either firstNeededLedger or
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupWork.cpp]: rework this comment: "applied" is a bit vague, what you're computing is the maximum ledger number that we're willing to do "bucket apply" on, in order to replay enough of transaction history.

File: src/catchup/CatchupWork.cpp [hunk 28]
```diff
@@ -0,0 +1,498 @@
...
+
+    // if we are about to apply buckets just after LCL, we can as well apply
+    // transactions
+    if (catchupStart.first && catchupStart.second <= lastClosedLedger + 1)
+    {
+        catchupStart = std::make_pair(false, catchupStart.second);
+    }
+
+    auto ledgerRange =
+        LedgerRange{std::min(configuration.toLedger(), catchupStart.second),
```
Comment (MonsieurNicolas) [inline src/catchup/CatchupWork.cpp]: when would we have `configuration.toLedger() >= catchupStart.second` ? and when it happens, why would we preserve the value of `catchupStart.first` ?

File: src/ledger/LedgerManagerImpl.cpp (hunk 26 above)
Comment (MonsieurNicolas) [reply src/ledger/LedgerManagerImpl.cpp]: this question still stands

Comment (vogel) [reply src/ledger/LedgerManagerImpl.cpp]: In previous version we were not storing state on bucket apply, because we were always applying buckets after that. Even in CATCHUP_MINIMAL case it was replaying buffered transactions (we always had at least one).

Now with commandline catchup case buffered transactions are always empty, so we need to store some state in database.

And that case is similar to applying buckets and then crashing during downloading transactions. Database will be in correct state.

**Comment by vogel** [comment]
All requested changes implemented - now code is a bit clearer.

File: src/ledger/LedgerManagerImpl.cpp (hunk 26 above)
Comment (MonsieurNicolas) [reply src/ledger/LedgerManagerImpl.cpp]: I see, thanks for the clarification & background on why this is needed. Maybe @jonjove will want to move this closer to the actual "apply buckets" code in his PR (and if invariants are executed in command line mode)

**Comment by MonsieurNicolas** [comment]
OK, this PR looks good now. Thank you.

I think you need a last round of rebase + commit squash before we get this one merged.

Before merging it I'll add an extra commit for the Visual Studio project files updates.

Given that #1307 seems to still need some work, it looks like this PR wins :)

**Comment by MonsieurNicolas** [comment]
r+ 6a8e02556867e61c44586b7951a5be5fd2a4669a
